src/frontend/src/expr/mod.rs

```diff
@@ -151,7 +153,53 @@ macro_rules! impl_has_variant {
 impl_has_variant! {InputRef, Literal, FunctionCall, AggCall, Subquery}
 
 impl ExprImpl {
-    // We need to traverse inside subqueries.
+    pub fn get_and_change_correlated_input_ref(&mut self) -> Vec<InputRef> {
+        struct Changer {
+            correlated_inputs: Vec<InputRef>,
+            depth: usize,
+            index: usize,
+        }
+        impl ExprChanger for Changer {
+            fn change_correlated_input_ref(
+                &mut self,
+                correlated_input_ref: &mut CorrelatedInputRef,
+            ) {
+                if correlated_input_ref.depth() >= self.depth {
+                    self.correlated_inputs.push(InputRef::new(
+                        correlated_input_ref.index(),
+                        correlated_input_ref.return_type(),
+                    ));
+                    correlated_input_ref.index = self.index;
+                    self.index += 1;
+                }
+            }
+
+            fn change_subquery(&mut self, subquery: &mut Subquery) {
+                use crate::binder::BoundSetExpr;
+
+                self.depth += 1;
+                match &mut subquery.query.body {
+                    BoundSetExpr::Select(select) => select
+                        .select_items
+                        .iter_mut()
+                        .chain(select.group_by.iter_mut())
+                        .chain(select.where_clause.iter_mut())
+                        .for_each(|expr| self.change_expr(expr)),
+                    BoundSetExpr::Values(_) => {}
+                }
+                self.depth -= 1;
+            }
+        }
+        let mut changer = Changer {
+            correlated_inputs: vec![],
+            depth: 1,
+            index: 0,
+        };
+        changer.change_expr(self);
+        changer.correlated_inputs
+    }
+
+    /// Used to check whether the expression has [`CorrelatedInputRef`].
```

Comment:
```suggestion
    /// Used to check whether the expression has [`CorrelatedInputRef`].
    // We need to traverse inside subqueries.
```
Keep this comment, which explains why it's not included in macro